File: app/src/main/java/com/qiaoqiao/core/camera/vision/VisionPresenter.java
package com.qiaoqiao.core.camera.vision;


import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.google.api.services.vision.v1.model.AnnotateImageResponse;
import com.google.api.services.vision.v1.model.BatchAnnotateImagesResponse;
import com.google.api.services.vision.v1.model.EntityAnnotation;
import com.google.api.services.vision.v1.model.WebDetection;
import com.google.api.services.vision.v1.model.WebEntity;
import com.qiaoqiao.core.camera.vision.bus.VisionEntityClickEvent;
import com.qiaoqiao.core.camera.vision.model.VisionEntity;
import com.qiaoqiao.core.confidence.ui.Confidence;
import com.qiaoqiao.repository.DsRepository;

import java.lang.ref.Reference;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import de.greenrobot.event.Subscribe;

import static com.qiaoqiao.app.PrefsKeys.DEFAULT_CONFIDENCE_IMAGE;
import static com.qiaoqiao.app.PrefsKeys.DEFAULT_CONFIDENCE_LABEL;
import static com.qiaoqiao.app.PrefsKeys.DEFAULT_CONFIDENCE_LOGO;
import static com.qiaoqiao.app.PrefsKeys.KEY_CONFIDENCE_IMAGE;
import static com.qiaoqiao.app.PrefsKeys.KEY_CONFIDENCE_LABEL;
import static com.qiaoqiao.app.PrefsKeys.KEY_CONFIDENCE_LOGO;

public final class VisionPresenter extends VisionContract.Presenter {
	private final @NonNull VisionContract.View mView;
	private final Reference<Context> mContextReference;

	//------------------------------------------------
	//Subscribes, event-handlers
	//------------------------------------------------

	/**
	 * Handler for {@link VisionEntityClickEvent}.
	 *
	 * @param e Event {@link VisionEntityClickEvent}.
	 */
	@SuppressWarnings("unused")
	@Subscribe
	public void onEvent(VisionEntityClickEvent e) {
		mView.showDetail(e.getEntity(), e.getTransitionView());

	}


	//------------------------------------------------


	@Inject
	VisionPresenter(@Nullable Context cxt, @NonNull VisionContract.View view, @NonNull DsRepository dsRepository) {
		super(dsRepository);
		mContextReference = new WeakReference<>(cxt);
		mView = view;
	}

	@Inject
	void onInjected() {
		mView.setPresenter(this);
	}

	@Override
	public void addResponseToScreen(@NonNull BatchAnnotateImagesResponse response) {
		if (mContextReference.get() == null) {
			return;
		}
		final List<AnnotateImageResponse> annotates = response.getResponses();
		List<VisionEntity> filterList = new ArrayList<>();
		if (annotates != null && annotates.size() > 0) {
			final AnnotateImageResponse annotateImage = annotates.get(0);
			if (annotateImage != null) {
				final List<EntityAnnotation> labelAnnotations = annotateImage.getLabelAnnotations();
				if (labelAnnotations != null && labelAnnotations.size() > 0) {
					for (EntityAnnotation annotation : labelAnnotations) {
						if (!TextUtils.isEmpty(annotation.getDescription()) && annotation.getScore() > Confidence.getValueOnly(mContextReference.get(),
						                                                                                                       KEY_CONFIDENCE_LABEL,
						                                                                                                       DEFAULT_CONFIDENCE_LABEL)) {
							filterList.add(new VisionEntity(annotation, "LABEL_DETECTION").setActivated(true));
						}
					}
				}

				final List<EntityAnnotation> landmarkAnnotations = annotateImage.getLandmarkAnnotations();
				if (landmarkAnnotations != null && landmarkAnnotations.size() > 0) {
					for (EntityAnnotation annotation : landmarkAnnotations) {
						if (!TextUtils.isEmpty(annotation.getDescription()) && annotation.getScore() > Confidence.getValueOnly(mContextReference.get(),
						                                                                                                       KEY_CONFIDENCE_IMAGE,
						                                                                                                       DEFAULT_CONFIDENCE_IMAGE)) {
							filterList.add(new VisionEntity(annotation, "LANDMARK_DETECTION").setActivated(true));
						}
					}
				}
				final List<EntityAnnotation> logoAnnotations = annotateImage.getLogoAnnotations();
				if (logoAnnotations != null && logoAnnotations.size() > 0) {
					for (EntityAnnotation annotation : logoAnnotations) {
						if (!TextUtils.isEmpty(annotation.getDescription()) && annotation.getScore() > Confidence.getValueOnly(mContextReference.get(), KEY_CONFIDENCE_LOGO,
						                                                                                                       DEFAULT_CONFIDENCE_LOGO)) {
							filterList.add(new VisionEntity(annotation, "LOGO_DETECTION").setActivated(true));
						}
					}
				}

				final WebDetection webDetection = annotateImage.getWebDetection();
				if (webDetection != null) {
					final List<WebEntity> webEntities = webDetection.getWebEntities();
					if (webEntities != null && webEntities.size() > 0) {
						for (WebEntity entity : webEntities) {
							if (!TextUtils.isEmpty(entity.getDescription()) && entity.getScore() > Confidence.createFromPrefs(mContextReference.get(), KEY_CONFIDENCE_IMAGE, DEFAULT_CONFIDENCE_IMAGE)
							                                                                                 .getValue()) {
								filterList.add(new VisionEntity(entity, "WEB_DETECTION").setActivated(true));
							}
						}
					}
				}
			}
		}
		mView.addEntities(filterList);
	}

	@Override
	public void setRefreshing(boolean refresh) {
		mView.setRefreshing(refresh);
	}

	@Override
	public void clear() {
		mView.clear();
	}
}

File: app/src/main/java/com/qiaoqiao/core/camera/vision/VisionPresenter.kt
package com.qiaoqiao.core.camera.vision

import android.content.Context
import android.text.TextUtils
import com.google.api.services.vision.v1.model.BatchAnnotateImagesResponse
import com.qiaoqiao.app.PrefsKeys.*
import com.qiaoqiao.core.camera.vision.bus.VisionEntityClickEvent
import com.qiaoqiao.core.camera.vision.model.VisionEntity
import com.qiaoqiao.core.confidence.ui.Confidence
import com.qiaoqiao.repository.DsRepository
import com.qiaoqiao.rx.Composer
import com.qiaoqiao.rx.safeList
import de.greenrobot.event.Subscribe
import io.reactivex.Flowable
import java.lang.ref.Reference
import java.lang.ref.WeakReference
import javax.inject.Inject

class VisionPresenter @Inject constructor(cxt: Context, val view: VisionContract.View, dsRepository: DsRepository) : VisionContract.Presenter(dsRepository) {
    private val contextReference: Reference<Context> = WeakReference<Context>(cxt)

    @Inject
    fun onInjected() {
        view.setPresenter(this)
    }

    @Subscribe
    fun onEvent(e: VisionEntityClickEvent) {
        view.showDetail(e.entity, e.transitionView)

    }

    override fun setRefreshing(refresh: Boolean) {
        view.setRefreshing(refresh)
    }

    override fun clear() {
        view.clear()
    }

    override fun addResponseToScreen(response: BatchAnnotateImagesResponse) {
        if (contextReference.get() == null || response.responses == null || response.responses.isEmpty() || response.responses[0] == null) {
            view.addEntities(arrayListOf<VisionEntity>())
        } else {
            Flowable.merge(
                    Flowable.fromIterable(safeList(response.responses[0].labelAnnotations)).filter({
                        !TextUtils.isEmpty(it.description) && it.score > Confidence.getValueOnly(contextReference.get() as Context,
                                KEY_CONFIDENCE_LABEL,
                                DEFAULT_CONFIDENCE_LABEL)
                    }).map({ VisionEntity(it, "LABEL_DETECTION").setActivated(true) }),
                    Flowable.fromIterable(safeList(response.responses[0].landmarkAnnotations)).filter({
                        !TextUtils.isEmpty(it.description) && it.score > Confidence.getValueOnly(contextReference.get() as Context,
                                KEY_CONFIDENCE_IMAGE,
                                DEFAULT_CONFIDENCE_IMAGE)
                    }).map({ VisionEntity(it, "LANDMARK_DETECTION").setActivated(true) }),
                    Flowable.fromIterable(safeList(response.responses[0].logoAnnotations)).filter({
                        !TextUtils.isEmpty(it.description) && it.score > Confidence.getValueOnly(contextReference.get() as Context, KEY_CONFIDENCE_LOGO,
                                DEFAULT_CONFIDENCE_LOGO)
                    }).map({ VisionEntity(it, "LOGO_DETECTION").setActivated(true) }),
                    Flowable.fromIterable(safeList(response.responses[0].webDetection.webEntities)).filter({
                        !TextUtils.isEmpty(it.description) && it.score > Confidence.createFromPrefs(contextReference.get() as Context, KEY_CONFIDENCE_IMAGE, DEFAULT_CONFIDENCE_IMAGE)
                                .value
                    }).map({ VisionEntity(it, "WEB_DETECTION").setActivated(true) })
            ).compose(Composer()).toList().subscribe({ it -> view.addEntities(it) })
        }
    }
}